### PR TITLE
Change lock failure to retryable exception

### PR DESCRIPTION
**Pull request by ohadzeliger**

### CODE REVIEW

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneLockFailureTest.java [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    protected static final Index COMPLEX_PARTITIONED = complexPartitionedIndex(Map.of(
+            IndexOptions.TEXT_TOKENIZER_NAME_OPTION, AllSuffixesTextTokenizer.NAME,
+            INDEX_PARTITION_BY_FIELD_NAME, "timestamp",
+            INDEX_PARTITION_HIGH_WATERMARK, "10"));
+
+    @Test
+    void testAddDocument() throws IOException {
+        try (final FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, SIMPLE_DOC, SIMPLE_INDEX);
+            grabLockExternally(SIMPLE_INDEX, context);
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneLockFailureTest.java:72]: I feel like we should grab the lock in a separate transaction. I can't think of any situation where that should matter, but it much more accurately represents real use cases.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneLockFailureTest.java [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+            rebuildIndexMetaData(context, COMPLEX_DOC, COMPLEX_PARTITIONED);
+            grabLockExternallyForPartition(COMPLEX_PARTITIONED, context, 1);
+            context.commit();
+        }
+
+        try (final FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, COMPLEX_DOC, COMPLEX_PARTITIONED);
+            // This fails since the repartition tries to take a lock
+            // The exception here is the RecordCore wrapper around the Lucene exception
+            Assertions.assertThrows(RecordCoreException.class, () ->
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneLockFailureTest.java:212]: `assertThrows` returns the exception, so you could assert that the cause is correct.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneLockFailureTest.java [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+        }
+    }
+
+    private void grabLockExternally(final Index index, final FDBRecordContext context) throws IOException {
+        final FDBDirectory directory = new FDBDirectory(recordStore.indexSubspace(index), context, index.getOptions());
+        directory.obtainLock(IndexWriter.WRITE_LOCK_NAME);
+    }
+
+    private void grabLockExternallyForPartition(final Index index, final FDBRecordContext context, int partition) throws IOException {
+        final Subspace partitionSubspace = recordStore.indexSubspace(index).subspace(Tuple.from(partition, LucenePartitioner.PARTITION_DATA_SUBSPACE).add(0));
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneLockFailureTest.java:240]: The partition in this case is `0`, you are putting the group in the subspace.
```suggestion
    private void grabLockExternallyForPartition(final Index index, final FDBRecordContext context, int group) throws IOException {
        final Subspace partitionSubspace = recordStore.indexSubspace(index).subspace(Tuple.from(group, LucenePartitioner.PARTITION_DATA_SUBSPACE).add(0));
```

**Review by ScottDugas — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBExceptions.java [hunk 4]
```diff
@@ -143,6 +143,16 @@ public FDBStoreTransactionConflictException(FDBException cause) {
         }
     }
 
+    /**
+     * Transaction failed due to a lock already taken.
```
[the remark this answers is not in the document]
Comment (ohadzeliger) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBExceptions.java:147]: Done, wanted to keep this generic to handle all kinds of locks.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneExceptions.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    /**
+     * Wrap the exception thrown by Lucene by a {@link RecordCoreException} that can be later interpreted by the higher levels.
+     * @param message the exception's message to use
+     * @param ex the exception thrown by Lucene
+     * @param additionalLogInfo (optional) additional log infos to add to the created exception
+     * @return the {@link RecordCoreException} that should be thrown
+     */
+    public static RecordCoreException wrapException(String message, Exception ex, Object... additionalLogInfo) {
+        Object[] logInfo = EMPTY_KEYS_AND_VALUES;
+        if (ex instanceof LoggableKeysAndValues) {
```
[the remark this answers is not in the document]
Comment (ohadzeliger) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneExceptions.java]: Sure. I modeled this after the FDBExceptions code, but looks much cleaner without the copying around. Done.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneExceptions.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    private static final Object[] EMPTY_KEYS_AND_VALUES = new Object[0];
+
+    /**
+     * Wrap the exception thrown by Lucene by a {@link RecordCoreException} that can be later interpreted by the higher levels.
+     * @param message the exception's message to use
+     * @param ex the exception thrown by Lucene
+     * @param additionalLogInfo (optional) additional log infos to add to the created exception
+     * @return the {@link RecordCoreException} that should be thrown
+     */
+    public static RecordCoreException wrapException(String message, Exception ex, Object... additionalLogInfo) {
```
[the remark this answers is not in the document]
Comment (ohadzeliger) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneExceptions.java]: Done.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java [hunk 7]
```diff
@@ -165,7 +173,7 @@ private void fileLockCheckNewLock(byte[] val, long nowMillis) {
             if (existingTimeStamp > (nowMillis - timeWindowMilliseconds) &&
                     existingTimeStamp < (nowMillis + timeWindowMilliseconds)) {
                 // Here: this lock is valid
-                throw new RecordCoreException("FileLock: Lock failed: already locked by another entity")
+                throw new FDBDirectoryLockException("FileLock: Lock failed: already locked by another entity")
```
[the remark this answers is not in the document]
Comment (ohadzeliger) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java:176]: I would have liked that but the throws clauses force our hands. The lambda-thrown exceptions have to be RuntimeExceptions (hence RecordCoreExceptions) while Lucene declares "throws IOException". So we have to throw a runtimeException at the lower level, wrap with IOException for Lucene and wrap again with RecordCoreException for the record layer API.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneLockFailureTest.java (hunk 1 above)
Comment (ohadzeliger) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneLockFailureTest.java:72]: Done

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneLockFailureTest.java (hunk 2 above)
Comment (ohadzeliger) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneLockFailureTest.java]: Done. Makes sense.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneLockFailureTest.java (hunk 3 above)
Comment (ohadzeliger) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneLockFailureTest.java]: Yes. Done. Good catch.

**Review by ScottDugas — APPROVED** [review]

**Pull request merged**